### PR TITLE
Refactoring

### PR DESCRIPTION
**Pull request by tlodderstedt**
- added text about the different representations (id token as vp and vp_token)
- made description of the claims parameter syntax more formal
- extended claims syntax to support credential types

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,8 +1,8 @@
-# OpenID Connect vp_token response parameter extension 
+# OpenID Connect extension for Verifiable Presentations
 
 ## Abstract
 
-This specification defines the additional OpenID Connect authentication response parameter `vp_token`. The new parameter allows OpenID Connect OPs to provide RPs with End-User claims as W3C Verifiable Presentations or W3C Verifiable Credentials in addition to claims provided in the`id_token` and/or via Userinfo responses.
+This specification defines an extension of OpenID Connect to allows OpenID Connect OPs to provide RPs with End-User claims as W3C Verifiable Presentations or W3C Verifiable Credentials in addition to claims provided in the `id_token` and/or via Userinfo responses.
```
Comment (AdamJLemmon) [inline README.md]: `to allows` => `to allow`

If we want to make it even perhaps a bit more generic and not throw the W3C language too overtly we could adopt some of the language around... allowing OPs  to provide RPs with sets of claims where they are able to cryptographically prove control over /  be authenticated as the controller of that set of claims.   Where options include ... W3C VPs etc

I think we're finding the above as a bit more relatable to OIDC audiences as opposed to immediately calling out the intent being to bring in VCs / VPs

File: README.md [hunk 2]
```diff
@@ -12,69 +12,73 @@ This specification defines the additional OpenID Connect authentication response
...
 
 ## Overview
-- RP requests `vp_token` by adding an additional destination `vp_token` in the `claims` request parameter. 
-- Authentication event information is conveyed via the id token while it's up to the RP to determine what (additional) claims are allocated to id_token and vp_token, respectively.
-- `vp_token` is provided in the same response as the `id_token`. Depending on the response type, this can be either the authentication response or the token response. 
-- If the vp_token is returned in the frontchannel, a hash (`vp_hash`) of `vp_token` must be included in `id_token`.
+
+This specifications introduces two mechanisms to provide Verifiable Presentations:
+
+* ID Token as Verififiable Presentation: An ID Token may contain a claim `vp` or `vc` as defined in [JWT proof format](https://www.w3.org/TR/vc-data-model/#json-web-token), i.e. it is a valid OpenID Connect ID Token and a Verifiable Presentation at the same time. Consequently, this mechanism utilizes (and supports) the JWT proof format only.
```
Comment (AdamJLemmon) [inline README.md]: Do we want to support both `vp` and `vc`?
`vc` would generally be what the Holder (OP) has acquired during a credential issuance flow from an issuer and thus signed by that issuer.

So for the Holder to present that to the RP here they would need to wrap it with the `vp` envelope I think? Without then there would be no proof of control over the `vc`.  I suppose the Holder could present self asserted claims in the form of a `vc` if they wanted though... where they are both the issuer and subject.. but I don't know if that was the intended use here?

File: README.md [hunk 3]
```diff
@@ -12,69 +12,73 @@ This specification defines the additional OpenID Connect authentication response
...
+
+* ID Token as Verififiable Presentation: An ID Token may contain a claim `vp` or `vc` as defined in [JWT proof format](https://www.w3.org/TR/vc-data-model/#json-web-token), i.e. it is a valid OpenID Connect ID Token and a Verifiable Presentation at the same time. Consequently, this mechanism utilizes (and supports) the JWT proof format only. 
+* dedicated VP Token: a Verifiable Presentation is provided in a separate artifact designated as VP Token. Such a token is provided to the RP in addition to an `id_token` in the `vp_token` parameter. VP Tokens support Verifiable Presentations in JSON-LD as well JWT format including all respective proof formats. The `vp_token` is provided in the same response as the `id_token`. Depending on the response type, this can be either the authentication response or the token response. Authentication event information is conveyed via the id token while it's up to the RP to determine what (additional) claims are allocated to id_token and vp_token, respectively. If the vp_token is returned in the frontchannel, a hash (`vp_hash`) of `vp_token` must be included in `id_token`.
```
Comment (AdamJLemmon) [inline README.md]: If `vp_token` and `id_token` and both requested, is `claims` required?
I am thinking as an OP how to determine which claims to put into the `vp_token` and which to put into the `id_token`.

Does this: `it's up to the RP to determine what (additional) claims are allocated to id_token and vp_token` imply something similar?

Comment (tlodderstedt) [reply README.md]: Do you mean the claims parameter?

File: README.md (hunk 2 above)
Comment (tlodderstedt) [reply README.md]: Very good point! I don’t think a VC alone is useful.
@awoie @Sakurann WDYT?

File: README.md (hunk 3 above)
Comment (AdamJLemmon) [reply README.md]: yes the claims parameter. I am thinking in order to explicitly determine which claims to be in which artifacts, `id_token` vs. `vp_token`?

File: README.md [hunk 4]
```diff
@@ -40,81 +189,133 @@ The following is a non-normative example of how an RP would use the `claims` par
...
     &state=af0ifjsldkj
       
 ```
+### Verifiable Presentation
+
+The ID Token contains a `vp` element with the Verifiable Credential data. 
+
+```json
+{
+   "iss":"https://book.itsourweb.org:3000/wallet/wallet.html",
```
Comment (tlodderstedt) [inline README.md:231]: @AdamJLemmon is this supposed to be a SIOP compliant id token? I'm asking since although it contains a "sub_jwk! the "iss" claim is not "https://self-isued.me". How is the RP supposed to validate the signature of this id token?

File: README.md (hunk 3 above)
Comment (tlodderstedt) [reply README.md]: that’s what the claims parameter is used for. Claims that shall go into the vp_tokens are requested underneath „vp_token“ in the claims parameter.

File: README.md (hunk 4 above)
Comment (AdamJLemmon) [reply README.md:231]: We have been advocating for expanding the `iss` claim for self signed tokens to not be constrained to self-issued.me.
Here that claim is the URL of the wallet which is returning the `id_token`.  
Our rationale was that the RP would detect the presence of `sub_jwk` and know the `id_token` was self signed instead of checking that the `iss` claim is self-issued. So if sub_jwk is present then validate the signature against that public key otherwise follow existing protocol.  

We could update this to be self-issued.me though if we wanted to present this as being compliant with OIDC CORE as it sits today?

File: README.md (hunk 3 above)
Comment (AdamJLemmon) [reply README.md]: Right right! I'm just mentioning should we make the presence of `claims` a MUST if both `id_token` and `vp_token` are present?

File: README.md (hunk 2 above)
Comment (AdamJLemmon) [reply README.md]: I would agreed :)
ping @awoie @Sakurann

File: README.md (hunk 4 above)
Comment (tlodderstedt) [reply README.md:231]: I understand your rationale and sympathies with them. self-issued.me is much too limited. I would rather prefer an issuer specific iss value and an explicit marker for self issued id tokens. The presence of the sub_jwk could serve as such a marker. 

However, I don't think this example is OIDC Core compliant. According to the OIDC Core https://openid.net/specs/openid-connect-core-1_0.html#IDTokenValidation the client shall use the keys provided by the issuer to check the signature.

File: README.md (hunk 3 above)
Comment (tlodderstedt) [reply README.md]: The presence of the claims parameter in the request is a pre-requisite for the presence of the vp_token in the response. Can you propose text?

File: README.md (hunk 2 above)
Comment (tlodderstedt) [reply README.md]: we discussed this point and came to the conclusion there are use cases where a VC alone can be provided, e.g. if the VC is a bearer credential. I updated the text accordingly.

Comment (Sakurann) [reply README.md]: @tlodderstedt I think we discussed an example of a bearer VC inside a VP signed by a holder. I think for OIDC, supporting VP is enough for the use-cases I know now. also see [https://bitbucket.org/openid/connect/issues/1205/indicating-support-for-vcs-siop#comment-60170872](https://bitbucket.org/openid/connect/issues/1205/indicating-support-for-vcs-siop#comment-60170872)

File: README.md [hunk 5]
```diff
@@ -224,43 +224,28 @@ The successful authentication response contains an `id_token` and `state`.
 ```
 ### Verifiable Presentation
 
-The ID Token contains a `vp` element with the Verifiable Credential data. 
+The ID Token contains a `VerifiablePresentation` element with the Verifiable Presentation and Verifiable Credential data.
```
Comment (tlodderstedt) [inline README.md]: @Sakurann 
The example as previously present in the PR was provided by Adam and was compatible with the W3C spec. Adam, Kim, Oliver and myself like the simplicity (id token as VP) and standard compliance of this representation.

We can certainly discuss your alternative proposal, but I think the best way would be a dedicated PR. 

May I ask you to reverse your commit and create a dedicated PR?

File: README.md (hunk 3 above)
Comment (AdamJLemmon) [reply README.md]: I don't mind your language there actually, but last question before landing on some language I think would be if we allow just `vp_token` as a response_type? and if so would we also require `claims` be used in that case?  Or we only require `claims` if both `vp_token` and `id_token` are present?

File: README.md (hunk 4 above)
Comment (AdamJLemmon) [reply README.md:231]: Right, so your suggestion for the time being in this doc would be to revert `iss` to self-issued.me for this example?

File: README.md (hunk 2 above)
Comment (awoie) [reply README.md]: to me, just using vp is fine.

File: README.md (hunk 4 above)
Comment (tlodderstedt) [reply README.md:231]: yes

**Review by tlodderstedt — COMMENTED** [review]
@AdamJLemmon let's come back to the question re an additional response type once we have discussed https://github.com/awoie/vp-token-spec/issues/16

File: README.md (hunk 3 above)
Comment (tlodderstedt) [reply README.md]: I suggest to postpone this discussion until we have discussed #16
